sdk/cosmos/azure-cosmos/azure/cosmos/aio/_asynchronous_availability_strategy_handler.py

```diff
@@ -0,0 +1,335 @@
+        :param request: The HTTP request to be executed
+        :type request: HttpRequest
+        :param execute_request_fn: Async function to execute the actual request
+        :type execute_request_fn: Callable[..., Awaitable[ResponseType]]
+        :returns: A tuple containing the response data and headers from the successful request
+        :rtype: ResponseType
+        :raises: Exception from the first request if all requests fail with transient errors
+        """
+        # Get available locations from global endpoint manager
+        available_locations = self._get_applicable_endpoints(request_params, global_endpoint_manager)
+        completion_status = AsyncHedgingCompletionStatus()
+
+        tasks = []
+        first_task: Optional[Task] = None
+        first_request_params_holder: SimpleNamespace = SimpleNamespace(request_params=None)
+        try:
+            # Create tasks for each location
+            for i in range(len(available_locations)):
+                task = asyncio.create_task(
+                    self.execute_single_request_with_delay(
+                        request_params,
+                        request,
+                        execute_request_fn,
+                        i,
+                        available_locations,
+                        completion_status,
+                        first_request_params_holder
+                    ))
+                tasks.append(task)
+                if i == 0:
+                    first_task = task
+
+            # Wait for tasks to complete
+            for completed_task in asyncio.as_completed(tasks):
+                try:
+                    result = await completed_task
+                    completion_status.set_completed()
+
+                    if completed_task is first_task:
+                        return result
+
+                    # successful response does not come from the initial request, record failure for it
+                    await self._record_cancel_for_first_request(first_request_params_holder, global_endpoint_manager)
+                    return result
+                except Exception as e: #pylint: disable=broad-exception-caught
+                    if completed_task is first_task:
+                        completion_status.set_completed()
+                        raise e
+                    if self._is_non_transient_error(e):
+                        completion_status.set_completed()
+                        await self._record_cancel_for_first_request(
+                            first_request_params_holder,
+                            global_endpoint_manager)
+                        raise e
+
+            # if we have reached here, it means all tasks completed_task but all failed with transient exceptions
+            # in this case, raise the exception from the first task
+            completion_status.set_completed()
+            assert first_task is not None
+            assert first_task.exception() is not None
```

Comment:
We shouldn't use `assert` statements in production code. I also can't see anywhere where these would be handled internally, and we wouldn't want any non-HTTP response exceptions to leak out from this implementation.

Reply:
Changed to raise RuntimeError